### PR TITLE
Ninja Blade and Skills

### PR DESCRIPTION
**Pull request by Kurfursten**
This is the last ninja buff for a while. We need to see how people adapt to these new abilities.

**Energy Katana** - Replaces the Katana
* Only stuck to your hand while it is deployed. Anyone can turn it on but only a ninja can turn it off.
* If you attack yourself with the Katana it conveniently examines for you, because its thin sprite is hard to shift click in the moment.
* Examine readout is much better than the weeb version. It shows you if teleport is active, remaining cooldown seconds, and cost in joules to refresh. Color coded.
* Toggle teleport is now an action button
* All the perks of an esword - notably hotness, wall/airlock cutting, and armor penetration.

**Darkvision** - Think Grey, not Vampire. It's only dim vision instead of pitch black.

**Under the Hood**
New daemon datums can flexibly add behaviors to two very different items.

🆑 
* rscadd: Ninjas are now equipped with the Energy Katana.
* rscadd: Ninjas will now automatically pick shuriken up by walking over them. If you have a stack of shurikens in hand it will add to the stack.
* rscadd: Ninjas can now see in the dark.

### CODE REVIEW

**Comment by spoonylove** [comment]
The spider clan grows stronger.

**Comment by SonixApache** [comment]
Really nice changes, Kurf.

**Comment by Kitfox88** [comment]
Does the katana still not conduct electricity, like the old one?

**Comment by Kurfursten** [comment]
Yes, unlike normal eswords

**Comment by SonixApache** [comment]
why do normal eswords conduct electricity 😒

**Comment by jknpj** [comment]
They don't.

**Comment by Kurfursten** [comment]
Don't they inherit siemens_coeffecient = 1 all the way up from /item/?

**Comment by Blithering** [comment]
Any chance shurikens could autothrow like bolas?

**Comment by Kurfursten** [comment]
That actually might make them too insane, but we can give it a try

**Comment by Tokanova** [comment]
can we get a prompt at start actually asking people signing up that will be a late join ninja if they want to be ninja? late join ninja is always a quick death sentence.

**Comment by Kurfursten** [comment]
discuss in its own issue #21613

**Comment by Kurfursten** [comment]
I actually played around with this with autothrow enabled (which is literally a 1 line change) and it just felt too good. Let's leave it for later if ninja really needs more QoL.

Meanwhile this is ready to go.

File: code/datums/gamemode/role/ninja.dm [hunk 1]
```diff
@@ -474,4 +504,44 @@
...
+/obj/item/weapon/katana/hesfast/suicide_act(mob/user)
+	to_chat(viewers(user), "<span class='danger'>[user] is slicing \his chest open with the [src.name]! It looks like \he's trying to commit sudoku.</span>")
+	return(SUICIDE_ACT_BRUTELOSS)
+
+/obj/item/weapon/katana/hesfast/preattack(atom/target, mob/user, proximity_flag, click_parameters)
+	if(target == user)
+		examine(user)
+		return 1
+	else
+		..()
```
Comment (Exxion) [inline code/datums/gamemode/role/ninja.dm]: Should be `return ..()`, shouldn't it?

File: code/datums/gamemode/role/ninja.dm [hunk 2]
```diff
@@ -355,63 +377,74 @@
...
+	var/cc = max(round((daemon.cooldown - world.time)/10),0)
+	to_chat(user,"<span class='notice'>The hilt displays its status in the form of a cryptic readout.</span>")
+	to_chat(user,"<span class='notice'>TP: </span>[daemon.active ? "<span class='good'>I":"<span class='warning'>O"]</span><span class='notice'>; CD: [cc ? "[cc]s ([cc*10*CHARGE_COST_MULTIPLIER]J)</span>" : "</span><span class='warning'><B>X</B></span>"]")
+
+/obj/item/weapon/melee/energy/sword/ninja/preattack(atom/target, mob/user, proximity_flag, click_parameters)
+	if(target == user)
+		examine(user)
+		return 1
+	else
+		..()
```
Comment (Exxion) [inline code/datums/gamemode/role/ninja.dm]: `return ..()`

File: code/datums/daemon.dm [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	if(cooldown > world.time)//you're trying to teleport when it's on cooldown.
+		return
+	var/mob/living/L = A
+	var/turf/SHHHHIIIING = get_step(L.loc, turn(L.dir, 180))
+	if(!SHHHHIIIING) //sanity for avoiding banishing our weebs into the shadow realm
+		return
+	cooldown = initial(cooldown) + world.time
+	if(telesound)
+		playsound(src, telesound,50,1)
+	user.forceMove(SHHHHIIIING)
```
Comment (Exxion) [inline code/datums/daemon.dm:43]: Not sure whether or not this is intentional, but you can teleport behind someone against a wall, right into the wall
If it's not intentional this should use `Move()` instead and should probably check the result of that to see if it's successful

File: code/datums/gamemode/role/ninja.dm [hunk 4]
```diff
@@ -82,6 +82,22 @@
...
+/obj/item/stack/shuriken/Crossed(atom/movable/A)
+	if(!ishuman(A))
+		return
+	var/mob/living/carbon/human/H = A
+	if(!isninja(H))
+		return
+	var/obj/item/stack/shuriken/S = locate(/obj/item/stack/shuriken) in H.held_items
+	if(S)
+		to_chat(H,"<span class='notice'>You add the shuriken to the stack.</span>")
+		S.amount++
```
Comment (Exxion) [inline code/datums/gamemode/role/ninja.dm]: Should increase by its own `amount` rather than always one

File: code/datums/gamemode/role/ninja.dm [hunk 5]
```diff
@@ -474,4 +504,44 @@
...
+		var/cooldowncalculated = round((daemon.cooldown - world.time)/10)
+		message += "Your steel has unleashed its dark and unwholesome power, so it's tapped out right now. It'll be ready again in [cooldowncalculated] seconds."
+	if(daemon.active)
+		message += " Your teleport is active, just like Goku's Shunkan Idou (Instant Transmission for Gaijin).</span>"
+	else
+		message += " Your teleport is inactive, just like a no-warp trap room in Aincrad.</span>"
+	to_chat(user, "[message]")
+
+/obj/item/weapon/katana/hesfast/suicide_act(mob/user)
+	to_chat(viewers(user), "<span class='danger'>[user] is slicing \his chest open with the [src.name]! It looks like \he's trying to commit sudoku.</span>")
```
Comment (Exxion) [inline code/datums/gamemode/role/ninja.dm]: This should probably be `visible_message()` but most of the other `suicide_act()`s do it this way so it's OK

File: code/datums/daemon.dm (hunk 3 above)
Comment (Kurfursten) [reply code/datums/daemon.dm:43]: I _think_ this is intentional

**Pull request merged**